File: lld/problems/category_b_machine_coding/incpp/in_memory_job_scheduler/models/job.cpp
#pragma once
#include<string>
#include<chrono>
#include"../util/utils.cpp"
using namespace std; 
enum class JobType {
    RECURRING, NON_RECURRING
};

enum class JobStatus {
    CREATED, SCHEDULED, SUCCESS, FAILED
};

class JobI {
    public: 
        virtual int getJobId() = 0; 
        virtual void setJobId(int id) = 0; 
        virtual string getData() = 0;
};
class ImmediateJob: public JobI {
    public: 
        int id; 
        string jobData; 
        JobType type; 
        chrono::seconds ts; 
    ImmediateJob(){}
    ImmediateJob(string jd, chrono::seconds t): jobData(jd), type(JobType::NON_RECURRING), ts(t) {}
    int getJobId()  {
        return this -> id; 
    }
    void setJobId(int id) {
        this -> id = id;
    }
    string getData(){
        return jobData;
    }
        
};


class ScheduleJob: public JobI {
    public: 
        int id; 
        string jobData; 
        JobType type; 
        int period; // in seconds;
        chrono::steady_clock::time_point nextTs; 

        ScheduleJob(){}
        ScheduleJob(string jd, int p): id(Counter::getCount()), jobData(jd), type(JobType::RECURRING), nextTs(chrono::steady_clock::now()+chrono::seconds(p)) {}
        int getJobId()  {
            return this -> id; 
        }
        void setJobId(int id) {
            this -> id = id;
        }
        string getData(){
            return jobData;
        }
        
};

class ExecutionJob {
    public: 
        int id; 
        JobI* job; 
        chrono::steady_clock::time_point ts; 
        JobStatus status; 
    ExecutionJob(){}
    ExecutionJob(JobI* jb, chrono::steady_clock::time_point t): id(Counter::getCount()), job(jb), ts(t), status(JobStatus::CREATED) {}
    int getJobId()  {
        return this -> id; 
    }
    void setJobId(int id) {
        this -> id = id;
    }
    string getData() {
        return this -> job -> getData();
    }
    void updateStatue(JobStatus status) {
        this -> status = status;
    }

        
};

File: lld/problems/category_b_machine_coding/ingolang/in_memory_job_scheduler_go/models/job.go
package models

import (
	"time"
	"in_memory_job_scheduler_go/util"
)

type JobType int

const (
	RECURRING JobType = iota
	NON_RECURRING
)

type JobStatus int

const (
	CREATED JobStatus = iota
	SCHEDULED
	SUCCESS
	FAILED
)

type JobI interface {
	GetJobId() int
	SetJobId(int)
	GetData() string
}

type ImmediateJob struct {
	Id      int
	JobData string
	Type    JobType
	Ts      time.Duration
}

func NewImmediateJob(jobData string, ts time.Duration) *ImmediateJob {
	return &ImmediateJob{
		Id:      util.GetCount(),
		JobData: jobData,
		Type:    NON_RECURRING,
		Ts:      ts,
	}
}

func (j *ImmediateJob) GetJobId() int { return j.Id }
func (j *ImmediateJob) SetJobId(id int) { j.Id = id }
func (j *ImmediateJob) GetData() string { return j.JobData }

type ScheduleJob struct {
	Id      int
	JobData string
	Type    JobType
	Period  int // in seconds
	NextTs  time.Time
}

func NewScheduleJob(jobData string, period int) *ScheduleJob {
	return &ScheduleJob{
		Id:      util.GetCount(),
		JobData: jobData,
		Type:    RECURRING,
		Period:  period,
		NextTs:  time.Now().Add(time.Duration(period) * time.Second),
	}
}

func (j *ScheduleJob) GetJobId() int { return j.Id }
func (j *ScheduleJob) SetJobId(id int) { j.Id = id }
func (j *ScheduleJob) GetData() string { return j.JobData }

type ExecutionJob struct {
	Id     int
	Job    JobI
	Ts     time.Time
	Status JobStatus
}

func NewExecutionJob(job JobI, ts time.Time) *ExecutionJob {
	return &ExecutionJob{
		Id:     util.GetCount(),
		Job:    job,
		Ts:     ts,
		Status: CREATED,
	}
}

func (j *ExecutionJob) GetJobId() int { return j.Id }
func (j *ExecutionJob) SetJobId(id int) { j.Id = id }
func (j *ExecutionJob) GetData() string { return j.Job.GetData() }
func (j *ExecutionJob) UpdateStatus(status JobStatus) { j.Status = status }

